### ISSUE TITLE
[caffe2] How to freeze a layer?

### ISSUE THREAD

**Issue by mks0601** [pytorch/pytorch#6402]
Hi, I`m working on caffe2.
I want to know how to freeze a certain layer, which means I don`t want to update a parameter (weight, bias) of a certain conv layer. How can I implement this feature? I found Stopgradient, Negategradient, Zerogradient in caffe2 doc, however none of them seems useful for me.

**Comment by Erotemic** [comment]
In pytorch the optimizer class (e.g. SGD, Adam) stores the learning rate for each layer. Set the learning rate of the layer you want to freeze to zero.

**Comment by mks0601** [comment]
Thanks for answering, however I'm not asking about pytorch, but caffe2.

**Comment by jimmyoic** [comment]
Maybe  try to delete the operators which update the weight?

**Comment by pradeephike** [comment]
@mks0601  Were you able to figure it out? I am also stuck on the same proble

**Comment by mks0601** [comment]
Not yet.

**Comment by jimmyoic** [comment]
How do you generate your training prototxt?
My first quick idea follows pytorch way @Erotemic mentioned is that you can set the input of lr as 0
For example
`op {
  input: "gpu_1/conv1_b_grad"
  input: "gpu_1/conv1_b_momentum"
  input: "gpu_1/lr"
  input: "gpu_1/conv1_b"
  output: "gpu_1/conv1_b_grad"
  output: "gpu_1/conv1_b_momentum"
  output: "gpu_1/conv1_b"
  name: ""
  type: "MomentumSGDUpdate"
  arg {
    name: "nesterov"
    i: 1
  }
  arg {
    name: "momentum"
    f: 0.899999976158
  }
  device_option {
    device_type: 1
    cuda_gpu_id: 1
  }
}`

This updates the bias of conv1 by MomentumSGDUpdate with learning rate as gpu_1/lr. We can replace this input by ZERO . I'm not sure is this the best way but I think it might work.

**Comment by xjtuljy** [comment]
@jimmyoic I didn't find any Python API to individually set the learning rate for specific layers -- it seems that all the layers are set with same learning rate in one GPU.
@mks0601 have you solve the problem?

Thanks.

**Comment by jimmyoic** [comment]
Have you tried not use the operator which updates the weight?
If you check your prototxt for training, there should be some operators used to update the weight, this is am example for updating the bias of my residual block 2:
`op {
  input: "gpu_0/block2_conv1_b_grad"
  input: "gpu_0/block2_conv1_b_momentum"
  input: "gpu_0/lr"
  input: "gpu_0/block2_conv1_b"
  output: "gpu_0/block2_conv1_b_grad"
  output: "gpu_0/block2_conv1_b_momentum"
  output: "gpu_0/block2_conv1_b"
  name: ""
  type: "MomentumSGDUpdate"
  arg {
    name: "nesterov"
    i: 1
  }
  arg {
    name: "momentum"
    f: 0.899999976158
  }
  device_option {
    device_type: 1
    cuda_gpu_id: 0
  }
}`
Though I haven't tried, I think disabling the operators updating the layers you want to freeze may work

**Comment by ecr23xx** [comment]
You can use operator `StopGradient`.

Read details in https://github.com/caffe2/caffe2/issues/1744.
